### ISSUE TITLE
Allow pointers to arbitrary data structures in Rust

### ISSUE THREAD

**Issue by janpaul123** [Zaplib/zaplib#125]
We should be able to send a `Box<T>` (mutable) or an `Arc<T>` (immutable) to JS. On the JS side you can't do anything with it, except send it back to Rust, or send it between Web Workers.

This is useful when doing complicated things in Rust, while the "locus of control" is still in Javascript.

The particular use case for Flux involves having a matrix that we don't need to read of modify directly on the JS side, so we want flexibility on how to represent it in Rust (e.g. swap out one matrix library for another).

**Comment by stevekrouse** [comment]
From the Rust side, it would look like this:

```rust
operation1() -> Box<T> {
  ... 
}

operation2(input: Box<T>) {
 ...
}
```

From the JS side it'd look like:

```js
const data = callRustSync('operation1', [])
callRustSync('operation2', [data])
```

**Comment by janpaul123** [comment]
Yeah exactly. To add some more details, we can call the type of `data` a `ZapPointer` perhaps, and we could even add a `zaplib.isZapPointer` in addition to our `zaplib.isZapBuffer`. (There are different ways to structure this so this is just one idea)

Re `how do we cast this into a Box<T>`, I would imagine doing `let data = params[0].as_box::<T>()`.

**Comment by janpaul123** [comment]
Workaround is something like this:

```rust
impl MyStruct {
    fn into_param(self) -> ZapParam {
        (Box::into_raw(Box::new(self)) as usize).to_string().into_param()
    }

    fn from_param(param: ZapParam) -> Self {
        *unsafe { Box::from_raw(param.as_str().parse::<usize>().unwrap() as *mut Self ) }
    }
}
```

Which has some overhead in the conversions from/to strings, plus it leaks memory if you drop the pointer on the JS side.
